### PR TITLE
[mlir][vector] Clarify the semantics of BroadcastOp

### PR DESCRIPTION
**Pull request by banach-space**
Clarifies the semantics of `vector.broadcast` in the context of scalable
vectors. In particular, broadcasting a unit scalable dim, `[1]`, is not
valid unless there's a match between the output and the input dims.
See the examples below for an illustration:

```mlir
// VALID
 %0 = vector.broadcast %arg0 : vector<[1]xf32> to vector<4x[1]xf32>
// INVALID
 %0 = vector.broadcast %arg0 : vector<[1]xf32> to vector<[4]xf32>
// VALID FIXED-WIDTH EQUIVALENT
 %0 = vector.broadcast %arg0 : vector<1xf32> to vector<4xf32>
```

Documentation, the Op verifier and tests are updated accordingly.

### CODE REVIEW

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 1]
```diff
@@ -2391,12 +2391,28 @@ mlir::vector::isBroadcastableTo(Type srcType, VectorType dstVectorType,
...
     int64_t dstDim = dstVectorType.getDimSize(lead + r);
-    if (srcDim != 1 && srcDim != dstDim) {
+    if ((srcDim != 1 && srcDim != dstDim))
+      mismatch = true;
+
+    // Check scalable flags
+    bool srcDimScalableFlag = srcVectorType.getScalableDims()[r];
+    bool dstDimScalableFlag = dstVectorType.getScalableDims()[lead + r];
+    if ((srcDim == 1 && srcDimScalableFlag && dstDim != 1) ||
+        (srcDimScalableFlag && !dstDimScalableFlag))
```
Comment (nujaa) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2406]: It got me thinking, what would be the expected behaviour of something like:
```
 %0 = vector.broadcast %arg0 : vector<nxf32> to vector<[n]xf32>
```
IMO it should not be supported as physically equivalent to a usecase
```
%1 = vector.broadcast %arg0 : vector<nxf32> to vector<vscale*nxf32>
```
Which is not invalid for fixed dimensions. Do you think this handles the cases ?
```suggestion
        (srcDimScalableFlag != dstDimScalableFlag))
```

**Review by nujaa — COMMENTED** [review]
Hi, I think there's a missed opportunity here.  😃

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.h [hunk 2]
```diff
@@ -68,9 +68,13 @@ enum class BroadcastableToResult {
   DimensionMismatch = 2,
   SourceTypeNotAVector = 3
 };
+struct VectorDim {
+  int64_t dim;
+  bool scalableFlag;
+};
```
[the remark this answers is not in the document]
Comment (banach-space) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.h:74]: This PR is unrelated to that discussion. I'm only adding this here to avoid adding new set of params to `isBroadcastableTo`.

I believe that before we commit to any new wider API, we should discuss the internal representation of `VectorType` and how scalable dimensions are represented. I am working on a proposal, but that's not yet ready to share 😅 I'm hoping to have something in the coming weeks.

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 1 above)
Comment (banach-space) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2406]: If you have e.g. `[2]` and `[4]` (i.e. `vscale * 2` and `vscale * 4`), then that's already "rejected" as "mismatching dims":
* https://github.com/llvm/llvm-project/blob/0dcada94bb1ae79f0edd91013038098c62a96b3b/mlir/lib/Dialect/Vector/IR/VectorOps.cpp#L2396

Is that the case you had in mind?

Comment (nujaa) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2406]: The case I pointed out was more src = `2`  and dest = `[2]`. srcDim == dstDim, so no mismatch on line 2399.  and we have `!srcDimScalableFlag`  so no mismatch on line 2406. Whereas I think this is wrong.
```
 %0 = vector.broadcast %arg0 : vector<2xf32> to vector<[2]xf32>
```

Comment (banach-space) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2406]: Ah, nice, great catch! In my head I had one case that wouldn't work with `!=`, but now I am failing to recall that 😂 

Let me send an update - thanks very much for pointing this out 🙏🏻

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.h [hunk 3]
```diff
@@ -68,9 +68,13 @@ enum class BroadcastableToResult {
   DimensionMismatch = 2,
   SourceTypeNotAVector = 3
 };
+struct VectorDim {
+  int64_t dim;
+  bool scalableFlag;
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.h:73]: nit: I don't find this name very descriptive, `true`/`false` doesn't have a clear mapping to static/scalable. Maybe something like `isScalable` would be more fitting?

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.h [hunk 4]
```diff
@@ -68,9 +68,13 @@ enum class BroadcastableToResult {
   DimensionMismatch = 2,
   SourceTypeNotAVector = 3
 };
+struct VectorDim {
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.h:71]:
```suggestion

struct VectorDim {
```

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 5]
```diff
@@ -2391,12 +2391,28 @@ mlir::vector::isBroadcastableTo(Type srcType, VectorType dstVectorType,
   // (all leading dimensions are simply duplicated).
   int64_t lead = dstRank - srcRank;
   for (int64_t r = 0; r < srcRank; ++r) {
+    bool mismatch = false;
+
+    // Check fixed-width dims
     int64_t srcDim = srcVectorType.getDimSize(r);
     int64_t dstDim = dstVectorType.getDimSize(lead + r);
-    if (srcDim != 1 && srcDim != dstDim) {
+    if ((srcDim != 1 && srcDim != dstDim))
```
Comment (kuhar) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2399]: These parens don't disambiguate anything

```suggestion
    if (srcDim != 1 && srcDim != dstDim)
```

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 6]
```diff
@@ -2391,12 +2391,28 @@ mlir::vector::isBroadcastableTo(Type srcType, VectorType dstVectorType,
   // (all leading dimensions are simply duplicated).
   int64_t lead = dstRank - srcRank;
   for (int64_t r = 0; r < srcRank; ++r) {
```
Comment (kuhar) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2393]: Consider renaming `r` to `rank`

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 7]
```diff
@@ -2391,12 +2391,28 @@ mlir::vector::isBroadcastableTo(Type srcType, VectorType dstVectorType,
...
+    bool mismatch = false;
```
Comment (kuhar) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2394]: How is `mismatch` different from `mismatchingDims`?

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 8]
```diff
@@ -2391,12 +2391,28 @@ mlir::vector::isBroadcastableTo(Type srcType, VectorType dstVectorType,
...
+
+    // Check fixed-width dims
```
Comment (kuhar) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2396]:
```suggestion
    // Check fixed-width dims.
```

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 9]
```diff
@@ -2391,12 +2391,28 @@ mlir::vector::isBroadcastableTo(Type srcType, VectorType dstVectorType,
...
     int64_t srcDim = srcVectorType.getDimSize(r);
     int64_t dstDim = dstVectorType.getDimSize(lead + r);
-    if (srcDim != 1 && srcDim != dstDim) {
+    if ((srcDim != 1 && srcDim != dstDim))
+      mismatch = true;
+
+    // Check scalable flags
```
Comment (kuhar) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2402]:
```suggestion
    // Check scalable flags.
```

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 10]
```diff
@@ -2406,16 +2422,25 @@ mlir::vector::isBroadcastableTo(Type srcType, VectorType dstVectorType,
...
   if (res == BroadcastableToResult::SourceRankHigher)
     return emitOpError("source rank higher than destination rank");
-  if (res == BroadcastableToResult::DimensionMismatch)
-    return emitOpError("dimension mismatch (")
-           << mismatchingDims.first << " vs. " << mismatchingDims.second << ")";
+  if (res == BroadcastableToResult::DimensionMismatch) {
+    std::string msg =
+        (Twine("dimension mismatch (") +
+         (mismatchingDims.first.scalableFlag ? "[" : "") +
+         std::to_string(mismatchingDims.first.dim) +
```
Comment (kuhar) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2436]: Doesn't this materialization defeat the purpose of  `Twine`? I'd think `Twine` should handle number to string conversion.

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 11]
```diff
@@ -367,6 +367,8 @@ def Vector_BroadcastOp :
                                s_1     x .. x s_j x .. x s_k
                <duplication>         <potential stretch>
        ```
+    * a scalable unit dimension, `[1]`, must match exactly.
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:370]: In isolation, I don't understand this requirement. Could you explain what it means for dimensions to match and/or provide an example?

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 6 above)
Comment (banach-space) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2393]: Note that `r` represents "dimension index" rather than rank. So, this should be something like `dimIdx` instead. Let me try that.

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 7 above)
Comment (banach-space) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2394]: Naming is hard ... Here's an attempt to explain:

1.  `mismatch` - bool to model whether a mismatch in dims has been found. I am introducing this to be able to replace logic like this:
```cpp
if (complex_conditions_for_fixed_width_widh_dims || complex_conditions_for_scalable_flags)
  // Do sth when dims mismatch
```

with something like this:
```cpp
if (complex_conditions_for_fixed_width_dims)
  mismatch = true;
  
if (complex_conditions_for_scalable_flags)
  mismatch = true;
  
if (mismatch)
  // Do sth when dims mismatch
```

2. `mismatchingDims` - pointer to the result. If `! = nullptr` then we'll use it to pass the first dimension that mismatched. I'll add `!=nullptr` to disambiguate a bit.

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td (hunk 11 above)
Comment (banach-space) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:370]: It always takes me a while to parse the previous two conditions. And this one is meant to be read together 😅 It feels that we should be able to phrase this better, but I'm failing to find a better description. 

Here are some examples:

**Example 1**
The following makes sense for fixed-width `1`, because there would be exactly 1 element to broadcast. Always. However, with scalable vectors (and scalable one `[1]`), there could be `1`, `2`, `4` ...  elements - we won't know the actual number until run-time.
```mlir
// This is fine:
%0 = vector.broadcast %arg0 : vector<1xf32> to vector<[4]xf32>
// This ambiguous and is disabled:
%1 = vector.broadcast %arg0 : vector<[1]xf32> to vector<[4]xf32>
```

Note, that the previous condition makes an exception for "fixed width" 1:
> the sizes in the trailing dimensions n-k < i <= n with j=i+k-n match exactly as s_j = t_i or s_j = 1:

The condition that I'm adding is meant to clarify that for `[1]` there is no exception.

**Example 2**
Again, the previous condition makes an exception for "fixed width" 1 and hence the following makes sense when using `1`. Scalable one, `[1]`, requires extra restrictions:
```mlir
// The trailing dims match exactly (`1 == 1`):
%0 = vector.broadcast %arg0 : vector<1xf32> to vector<4x1xf32>
// The trailing dims match exactly (`[1] == [1]`)
%1 = vector.broadcast %arg0 : vector<[1]xf32> to vector<4x[1]xf32>
// The trailing dims do not match exactly (`1 != [1]`)
%2 = vector.broadcast %arg0 : vector<1xf32> to vector<4x[1]xf32>
```

**Tl;Dr**

If there's exactly 1 element then we can obviously broadcast that to whatever size we need. However, `[1]` doesn't not qualify as "exactly 1" :)

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 7 above)
Comment (kuhar) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2394]: Oh, I see what you mean now. I think having a comment with a similar explanation would help here.

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 12]
```diff
@@ -2406,16 +2422,24 @@ mlir::vector::isBroadcastableTo(Type srcType, VectorType dstVectorType,
...
     return success();
   if (res == BroadcastableToResult::SourceRankHigher)
     return emitOpError("source rank higher than destination rank");
-  if (res == BroadcastableToResult::DimensionMismatch)
-    return emitOpError("dimension mismatch (")
-           << mismatchingDims.first << " vs. " << mismatchingDims.second << ")";
+  if (res == BroadcastableToResult::DimensionMismatch) {
+    std::string msg = (Twine("dimension mismatch (") +
+                       (mismatchingDims.first.isScalable ? "[" : "") +
+                       std::to_string(mismatchingDims.first.dim) +
```
Comment (kuhar) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2435]: Doesn't `Twine` allow for concatenating numbers directly? I'd think that we can construct a `Twine` out of a number and then add it to the previous `Twine`

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp [hunk 13]
```diff
@@ -2390,13 +2390,29 @@ mlir::vector::isBroadcastableTo(Type srcType, VectorType dstVectorType,
...
+
+    // Check scalable flags.
+    bool srcDimScalableFlag = srcVectorType.getScalableDims()[dimIdx];
+    bool dstDimScalableFlag = dstVectorType.getScalableDims()[lead + dimIdx];
+    if ((srcDim == 1 && srcDimScalableFlag && dstDim != 1) ||
+        (srcDimScalableFlag != dstDimScalableFlag))
+      mismatch = true;
+
+    if (mismatch) {
+      if (mismatchingDims != nullptr) {
```
Comment (kuhar) [inline mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2410]: This helps, thanks!

File: mlir/lib/Dialect/Vector/IR/VectorOps.cpp (hunk 12 above)
Comment (banach-space) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp:2435]: > Doesn't Twine allow for concatenating numbers directly?

Nope (but I might be missing sth): https://github.com/llvm/llvm-project/blob/b1234ddbe2652aa7948242a57107ca7ab12fd2f8/llvm/include/llvm/ADT/Twine.h#L554-L575

However, you can create a `Twine` from `unsigned`: https://github.com/llvm/llvm-project/blob/b1234ddbe2652aa7948242a57107ca7ab12fd2f8/llvm/include/llvm/ADT/Twine.h#L349

Let me do that instead.

Comment (kuhar) [reply mlir/lib/Dialect/Vector/IR/VectorOps.cpp]: Or alternatively, why not use the `<<` operator on the diagnostic itself? `return emitOpError("Foo") << bar << "baz";

**Comment by banach-space** [comment]
Ping @kuhar :)

**Review by kuhar — APPROVED** [review]

**Comment by banach-space** [comment]
@nujaa Any other suggestions from you or shall I land it?
